### DISCUSSION TITLE
Where should I put the app(Navigation::class)...

### DISCUSSION THREAD

**Discussion by MathieuBujold** [spatie/laravel-navigation#11] [General] [answered]
Hi
This look very good.
Any idea where should I put the app(Navigation::class)... with all of my tree sctructure.

A very simple Tutorial would be very nice.

Mathieu

**Comment by freekmurze** [comment, marked as answer]
Typically, you would add this in a service provider.

**Reply by MathieuBujold** [reply]
I tried this without success. I will wait. 
Someone with more knowledge will at some point release a tutorial. Thank you for your response.

I include Spatie\Navigation\NavigationServiceProvider::class, in my config/app.php and create a shortcut but I still have error.
This look Awesome! Can wait to figure this out.

**Reply by MathieuBujold** [reply]
It's like they don't see my route. I always get a : Route [dashboard] not defined.

**Comment by jameswagoner** [comment]
You need to build your navigation inside of `$this->app->booted()` callback:

```php
public function boot() {
    $this->app->booted(function () {
        app(Navigation::class)->add(...);
    });
}
```

Alternatively, you could build your navigation in a middleware.  This was my approach since I am using Inertia and used the `HandleInertiaRequests` middleware.

```php
public function share(Request $request)
{
    return array_merge(parent::share($request), [
        'navigation' => $this->buildNavigation()->tree()
    ]
}

protected function buildNavigation()
{
    return app(Navigation::class)->add(...);
}
```

**Comment by OctaneInteractive** [comment]
Hi @jameswagoner, any invocation of `route()` triggers an error, regardless of the code pattern.

I've so far failed to get this package to work as a Service Provider, and given the problems other people have had, it seems to me the README would benefit from a quick edit.

**Comment by jameswagoner** [comment]
Yeah, I did experience that as well. I felt it was kinda silly do have the navigation built in essentially a callback of callback in a service provider. I was in L8.x at that time though.

I have a recent L10 and Inertia project started and getting to the same point with needing some navigation. I can follow up if I take a different approach.
